design.md

```diff
@@ -0,0 +1,121 @@
+
+The server will enforce TLS version 1.3 for maximum security. The server will prefer the TLS13-AES-256-GCM-SHA384 cipher suite to have the best possible encryption standard while maintaining efficient performance. 
+
+### Scoping Requests
+In order to scope requests to each client we will create a user for each client, and that will be authenticated against the user in the certificate sent from the client. The user on the host will match the first part of the email, before the `@`, which will be compared against the email we receive from the client. 
+
+The client will hard code this for now. Ideally this would be set via a configuration file. 
+
+#### Future Improvements
+* Use a certificate provider for the CA instead of a locally generating one
+* Utilize client certificate details to implement more fine grained access
+* Implement a dedicated secret management solution for private keys, such as Vault or AWS Secret Manager (or any other secure key store)
+* Automate certificate management going forward to remove human error and allow for easier revocation and rotation
+* Add monitoring for certificate expiration and authentication failures
+
+### Library
+The library will primarily utilize the builtin os package in Go to interact with the Linux operating system and accomplish the actual job management functions.
+
+The library will primarily utilize the os/exec builtin, with some use of the os/user and base level os builtin packages.
+
+### Output streaming
+In order to reduce overhead each command will pipe output to a file named with the a unique identifier tied to the process. This data will be read from the file when requested. We will check to see if the process is still running to determine whether or not to continually read from and stream data from the file, or serve all the output data at once to the client. 
+
+In the cases where processes are still running we will utilize server side streaming from gRPC to serve this data continually to the client until they close the connection or the process finishes.
+
+#### Future Improvements
+* Store output somewhere other than locally on the server, like S3
+* Look into using an external lib that enables tailing from an external source
+* Utilize job management to allow suspending and resuming jobs instead of only being able to start and terminate them
+
+## CLI
+The CLI will utilize the [cobra](https://github.com/spf13/cobra) CLI library.
+
+### Run a process
+The basic structure of the run command will be `run <binary_name> [<args>]`
+
+An example of running a process: `cmdctl run ls -la`
+
+### Get process info
+The basic structure of the get-status and get-output retrieval commands will be: `get-status/get-output [<unique_process_id>]`
+
+An example of querying for process status: `cmdctl get-status shawon-ls-idhhyqm`
```

Comment:
How are you generating this `shawon-ls-idshhyqm` ID?

Reply:
I can expand on this more in the block above, but essentially this will be generated from the name sent in the client cert, the command being run, and a randomly generated string.